File: src/main/java/com/brainstormer/config/prod/ProdSecurityConfig.java
package com.brainstormer.config.prod;

import com.brainstormer.service.CustomUserDetailsService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.security.SecurityProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.core.annotation.Order;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.crypto.password.PasswordEncoder;

@Profile("prod")
@Configuration
@EnableWebSecurity
// This is needed to get authentication to work
// Not 100% sure why, might want to look into spring docs
@Order(SecurityProperties.ACCESS_OVERRIDE_ORDER)
public class ProdSecurityConfig extends WebSecurityConfigurerAdapter {

    @Autowired
    private CustomUserDetailsService userDetailsService;
    @Autowired
    private PasswordEncoder passwordEncoder;

    @Override
    protected void configure(HttpSecurity httpSecurity) throws Exception {
        httpSecurity.csrf().disable();
        httpSecurity
                .authorizeRequests().antMatchers("/").permitAll()
                .and()
                .headers().frameOptions().sameOrigin()
                .and()
                .formLogin().disable()
                .logout().logoutUrl("/logout"); // TODO: this doesn't actuallly work, need to manually implement
                 // look here for how to do it: https://stackoverflow.com/questions/21987589/spring-security-how-to-log-out-user-revoke-oauth2-token
    }

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        auth.authenticationProvider(authenticationProvider()).userDetailsService(userDetailsService);
    }

    @Override
    @Bean
    public AuthenticationManager authenticationManagerBean() throws Exception {
        return super.authenticationManagerBean();
    }

    // This is needed to have proper reading from the database for initial authentication
    @Bean
    public DaoAuthenticationProvider authenticationProvider() {
        final DaoAuthenticationProvider authenticationProvider = new DaoAuthenticationProvider();
        authenticationProvider.setUserDetailsService(userDetailsService);
        authenticationProvider.setPasswordEncoder(passwordEncoder);
        return authenticationProvider;
    }
}

File: src/main/kotlin/spring/starter/config/prod/ProdSecurityConfig.kt
package spring.starter.config.prod

import spring.starter.service.CustomUserDetailsService
import org.springframework.boot.autoconfigure.security.SecurityProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.core.annotation.Order
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.crypto.password.PasswordEncoder

@Profile("prod")
@Configuration
@EnableWebSecurity
// This is needed to get authentication to work
// Not 100% sure why, might want to look into spring docs
@Order(SecurityProperties.ACCESS_OVERRIDE_ORDER)
class ProdSecurityConfig(
        private var userDetailsService: CustomUserDetailsService,
        private var passwordEncoder: PasswordEncoder
) : WebSecurityConfigurerAdapter() {

    @Throws(exceptionClasses = Exception::class)
    override fun configure(httpSecurity: HttpSecurity) {
        httpSecurity.csrf().disable()

        httpSecurity
                .authorizeRequests().antMatchers("/").permitAll()
                .and()
                .headers().frameOptions().sameOrigin()
                .and()
                .formLogin().disable()
    }

    @Throws(exceptionClasses = Exception::class)
    override fun configure(auth: AuthenticationManagerBuilder) {
        auth.authenticationProvider(authenticationProvider()).userDetailsService(userDetailsService)
    }

    @Bean
    @Throws(exceptionClasses = Exception::class)
    override fun authenticationManagerBean(): AuthenticationManager {
        return super.authenticationManagerBean()
    }

    // This is needed to have proper reading from the database for initial authentication
    @Bean
    fun authenticationProvider(): DaoAuthenticationProvider {
        val authenticationProvider: DaoAuthenticationProvider = DaoAuthenticationProvider()
        authenticationProvider.setUserDetailsService(userDetailsService)
        authenticationProvider.setPasswordEncoder(passwordEncoder)
        return authenticationProvider
    }
}